app/services/expired_registration_cleanup_service.rb

```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+class ExpiredRegistrationCleanupService < ::WasteExemptionsEngine::BaseService
+  def run
+    return unless registration_ids.any?
+
+    ActiveRecord::Base.transaction do
+      # paper_trail will try creating a record for the deletion
+      # unless we execute within this block
+      PaperTrail.request(enabled: false) do
+        clear_registrations!
+      end
+      clear_versions!
+      clear_versions_archive!
+    end
+  end
+
+  private
+
+  # registration exemptions that were deregistered (ceased or revoked)
+  # or expired over 7 years ago
+  def registration_ids
+    @registration_ids ||=
+      WasteExemptionsEngine::RegistrationExemption
+      .where(
+        <<~SQL
+          deregistered_at <= '#{date}'
+          OR (state = 'expired' AND updated_at <= '#{date}')
+        SQL
+      )
+      .pluck(:registration_id)
+      .uniq
```

Comment:
Might need to also apply a limit to the number returned here, in order to protect service availability - see further comment below.